### PR TITLE
T74 create page login

### PR DESCRIPTION
**Pull request by xgarciagSopra**

### CODE REVIEW

File: src/app/components/login/login.component.html [hunk 1]
```diff
@@ -2,27 +2,74 @@
...
-        <label>Usuario:</label>
+      <mat-form-field class="example-full-width">
+        <mat-label>Usuario</mat-label>
         <input
           type="text"
-          class="form-control"
-          id="usuario"
-          placeholder="Ingresa tu usuario"
+          matInput
+          [formControl]="userFormControl"
```
Comment (ddiezgsop) [inline src/app/components/login/login.component.html]:
```suggestion
          [formControlName]="user"
```

File: src/app/components/login/login.component.html [hunk 2]
```diff
@@ -2,27 +2,74 @@
...
-      <div class="form-group mt-3">
-        <label>Contraseña:</label>
+        <mat-hint>Usuario valido</mat-hint>
+        <mat-error
+          *ngIf="
+            userFormControl.hasError('minlength') &&
+            userFormControl.hasError('pattern') &&
+            !userFormControl.hasError('required')
+          "
+        >
```
Comment (ddiezgsop) [inline src/app/components/login/login.component.html]:
```suggestion
        <mat-error
          *ngIf="showUserError()"
        >
```
Extract this logic to a function in the controller

File: src/app/components/login/login.component.html [hunk 3]
```diff
@@ -2,27 +2,74 @@
...
+        </mat-error>
+        <mat-error *ngIf="userFormControl.hasError('required')">
+          Usuario <strong> requerido</strong>
+        </mat-error>
+        <mat-error *ngIf="userFormControl.hasError('minlength')">
+          Usuario demasido <strong>corto</strong>
+        </mat-error>
+        <mat-error *ngIf="userFormControl.hasError('pattern')">
+          El usuario no permite <strong>caracteres especiales</strong>
+        </mat-error>
```
Comment (ddiezgsop) [inline src/app/components/login/login.component.html]: This is the same code, repeated 3 times, refactor into one `<mat-error>` with dinamic text

File: src/app/components/login/login.component.html [hunk 4]
```diff
@@ -2,27 +2,74 @@
...
+
+      <mat-form-field class="example-full-width">
+        <mat-label>Contraseña</mat-label>
         <input
           type="password"
-          class="form-control"
-          id="contrasena"
-          placeholder="Ingresa tu contraseña"
+          matInput
+          [formControl]="passwordFormControl"
```
Comment (ddiezgsop) [inline src/app/components/login/login.component.html]:
```suggestion
          [formControlName]="password"
```

File: src/app/components/login/login.component.html [hunk 5]
```diff
@@ -2,27 +2,74 @@
...
+          [errorStateMatcher]="matcher"
+          placeholder="Contraseña entre 4 y 12 carácteres"
         />
+        <mat-hint>Contraseña válida</mat-hint>
+        <mat-error
+          *ngIf="
+            passwordFormControl.hasError('minlength') &&
+            passwordFormControl.hasError('maxlength') &&
+            !passwordFormControl.hasError('required')
+          "
```
Comment (ddiezgsop) [inline src/app/components/login/login.component.html]:
```suggestion
        <mat-error
          *ngIf="showPasswordError()"
```

File: src/app/components/login/login.component.html [hunk 6]
```diff
@@ -2,27 +2,74 @@
...
+        </mat-error>
+        <mat-error *ngIf="passwordFormControl.hasError('required')">
+          Contraseña es <strong>requerida</strong>
+        </mat-error>
+        <mat-error *ngIf="passwordFormControl.hasError('minlength')">
+          Contraseña es demasiado <strong>corta</strong>
+        </mat-error>
+        <mat-error *ngIf="passwordFormControl.hasError('maxlength')">
+          Contraseña es demasiado <strong>larga</strong>
+        </mat-error>
```
Comment (ddiezgsop) [inline src/app/components/login/login.component.html]: Same as above

File: src/app/components/login/login.component.ts [hunk 7]
```diff
@@ -1,13 +1,64 @@
...
+    const isSubmitted = form && form.submitted;
+    return !!(
+      control &&
+      control.invalid &&
+      (control.dirty || control.touched || isSubmitted)
+    );
+  }
+}
+
+const pattern = '^[a-zA-Z0-9]*$';
```
Comment (ddiezgsop) [inline src/app/components/login/login.component.ts]: Rename the pattern to a more clear naming

File: src/app/components/login/login.component.ts [hunk 8]
```diff
@@ -1,13 +1,64 @@
...
+  isErrorState(
+    control: FormControl | null,
+    form: FormGroupDirective | NgForm | null
+  ): boolean {
+    const isSubmitted = form && form.submitted;
+    return !!(
+      control &&
+      control.invalid &&
+      (control.dirty || control.touched || isSubmitted)
+    );
```
Comment (ddiezgsop) [inline src/app/components/login/login.component.ts]:
```suggestion
    return 
      control &&
      control.invalid &&
      (control.dirty || control.touched || isSubmitted)
    ;
```

Once u have boolean operators, u do not need to have the casting value (!!)

File: src/app/components/login/login.component.ts [hunk 9]
```diff
@@ -1,13 +1,64 @@
...
+
+const pattern = '^[a-zA-Z0-9]*$';
+
 @Component({
   selector: 'login',
   templateUrl: './login.component.html',
   styleUrls: ['./login.component.scss'],
 })
 export class LoginComponent {
+  userFormControl = new FormControl('', [
```
Comment (ddiezgsop) [inline src/app/components/login/login.component.ts]:
```suggestion
  user = new FormControl('', [
```

File: src/app/components/login/login.component.ts [hunk 10]
```diff
@@ -1,13 +1,64 @@
...
+    Validators.required,
+    Validators.minLength(4),
+    Validators.pattern(pattern),
+  ]);
+  passwordFormControl = new FormControl('', [
```
Comment (ddiezgsop) [inline src/app/components/login/login.component.ts]:
```suggestion
  password = new FormControl('', [
```

File: src/app/components/login/login.component.ts [hunk 11]
```diff
@@ -1,13 +1,64 @@
...
+      !this.userFormControl.value ||
+      this.userFormControl.value.trim().length === 0 ||
+      !this.passwordFormControl.value ||
+      this.passwordFormControl.value.trim().length === 0 ||
+      this.passwordFormControl.hasError('required') ||
+      this.passwordFormControl.hasError('minlength') ||
+      this.passwordFormControl.hasError('maxlenght') ||
+      this.userFormControl.hasError('required') ||
+      this.userFormControl.hasError('minlength') ||
+      this.userFormControl.hasError('pattern')
```
Comment (ddiezgsop) [inline src/app/components/login/login.component.ts]: This have an infernal logic clausure. There are more options in the Forms that u can use to implement this

**Review by ddiezgsop — CHANGES_REQUESTED** [review]

File: src/app/components/login/login.component.ts (hunk 8 above)
Comment (juanjovSopra) [reply src/app/components/login/login.component.ts]: Tengo que ponerlo así porque control es formcontrol o nulo.

File: src/app/components/login/login.component.ts [hunk 12]
```diff
@@ -15,50 +16,65 @@ export class MyErrorStateMatcher implements ErrorStateMatcher {
...
+  }
+
+  getUserError() {
+    let user = this.form.controls.user;
+    if (user.errors!['required']) return 'Usuario requerido';
+    if (user.errors!['minlength']) return 'Usuario demasiado corto';
+    if (user.errors!['pattern'])
+      return 'El usuario no permite caracteres especiales';
+    else return 'Por favor inserta un usuario válido';
+  }
```
Comment (ddiezgsop) [inline src/app/components/login/login.component.ts]:
```suggestion
  getUserError(): string {
    const user = this.form.controls.user;
    if (user.errors['required']) return 'Usuario requerido';
    if (user.errors['minlength']) return 'El Usuario debe tener mínimo N caracteres';
    if (user.errors['pattern'])
      return 'El usuario no permite caracteres especiales';
    return 'Por favor inserta un usuario válido';
  }
```

File: src/app/components/login/login.component.ts [hunk 13]
```diff
@@ -15,50 +16,65 @@ export class MyErrorStateMatcher implements ErrorStateMatcher {
...
+
+  getPasswordError() {
+    let password = this.form.controls.password;
+    if (password.errors!['required']) return 'Contraseña requerida';
+    if (password.errors!['minlength']) return 'Contraseña demasiado corta';
+    if (password.errors!['maxlength']) return 'Contraseña demasiado larga';
+    else return 'Por favor inserta una ontraseña válida';
+  }
```
Comment (ddiezgsop) [inline src/app/components/login/login.component.ts]:
```suggestion
  getPasswordError(): string {
    const password = this.form.controls.password;
    if (password.errors['required']) return 'Contraseña requerida';
    if (password.errors['minlength']) return 'La contraseña debe tener mínimo N caracteres';
    if (password.errors['maxlength']) return 'La contraseña debe tener máximo N caracteres';
    return 'Por favor inserta una contraseña válida';
  }
```

File: src/app/components/login/login.component.ts [hunk 14]
```diff
@@ -15,50 +16,65 @@ export class MyErrorStateMatcher implements ErrorStateMatcher {
...
-      !this.passwordFormControl.value ||
-      this.passwordFormControl.value.trim().length === 0 ||
-      this.passwordFormControl.hasError('required') ||
-      this.passwordFormControl.hasError('minlength') ||
-      this.passwordFormControl.hasError('maxlenght') ||
-      this.userFormControl.hasError('required') ||
-      this.userFormControl.hasError('minlength') ||
-      this.userFormControl.hasError('pattern')
-    );
+    return this.form.valid;
```
Comment (ddiezgsop) [inline src/app/components/login/login.component.ts]: Are u using this function?

File: src/app/components/login/login.component.ts [hunk 15]
```diff
@@ -15,50 +16,65 @@ export class MyErrorStateMatcher implements ErrorStateMatcher {
...
-      control &&
-      control.invalid &&
+      control?.invalid &&
       (control.dirty || control.touched || isSubmitted)
     );
   }
 }
 
-const pattern = '^[a-zA-Z0-9]*$';
+const validCharacters = '^[a-zA-Z0-9]*$';
```
Comment (ddiezgsop) [inline src/app/components/login/login.component.ts]:
```suggestion
const alphaNumeric = '^[a-zA-Z0-9]*$';
```

Look for the custom Angular Validators if u have this already, to save u from declaring the pattern

**Review by ddiezgsop — CHANGES_REQUESTED** [review]

File: src/app/components/login/login.component.ts (hunk 13 above)
Comment (juanjovSopra) [reply src/app/components/login/login.component.ts]: si le quito las exclamaciones me salen errores

File: src/app/helpers/errorStateMatcher.ts [hunk 16]
```diff
@@ -0,0 +1,15 @@
+import { FormControl, FormGroupDirective, NgForm } from '@angular/forms';
+import { ErrorStateMatcher } from '@angular/material/core';
+
+export class MyErrorStateMatcher implements ErrorStateMatcher {
```
Comment (ddiezgsop) [inline src/app/helpers/errorStateMatcher.ts]:
```suggestion
export class ErrorMatcher implements ErrorStateMatcher {
```

**Review by ddiezgsop — APPROVED** [review]

**Pull request merged**